src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -128,17 +141,33 @@ void mir::frontend::XdgDecorationManagerV1::get_toplevel_decoration(wl_resource*
 }
 
 mir::frontend::XdgToplevelDecorationV1::XdgToplevelDecorationV1(
-    wl_resource* id, mir::frontend::XdgToplevelStable* toplevel) :
+    wl_resource* id,
+    mir::frontend::XdgToplevelStable* toplevel,
+    std::shared_ptr<mir::XdgDecorationsInterface> xdg_decorations) :
     wayland::XdgToplevelDecorationV1{id, Version<1>{}},
-    toplevel{toplevel}
+    toplevel{toplevel},
+    xdg_decorations{xdg_decorations}
 {
 }
 
 void mir::frontend::XdgToplevelDecorationV1::update_mode(uint32_t new_mode)
 {
+    switch (xdg_decorations->mode())
+    {
+    case ServerMode::always_ssd:
+        mode = Mode::server_side;
+        break;
+    case ServerMode::always_csd:
+        mode = Mode::client_side;
+        break;
+    case ServerMode::prefer_ssd:
+    case ServerMode::prefer_csd:
+        mode = new_mode;
+        break;
+    }
```

Comment:
There is no need for a switch anywhere. Your `XdgDecorationInterface` is doing the wrong job. What is needed is a Strategy interface (which can be implemented by `Decorations::Self`:

```c++
enum class DecorationsType { ssd, csd };

class DecorationStrategy
{
public:
    virtual auto default_style() const -> DecorationsType = 0;
    virtual auto request_style(DecorationsType requested) const -> DecorationsType = 0;

   ...
};
```

There can be a subclass of Decoration::Self in each of the Decoration named constructors that hard codes the correct responses